File: go/n_queens.go
package leetcode

import (
	"strings"
)

func solveNQueens(n int) [][]string {
	res := make([][]string, 0)
	backtrack(make(map[int]bool), make(map[int]bool), make(map[int]bool), n, make([]int, 0), &res)
	return res
}

func backtrack(rows map[int]bool, diags map[int]bool, antiDiags map[int]bool, n int, partialRes []int, res *[][]string) {
	if len(partialRes) >= n {
		*res = append(*res, writeOutNQueens(partialRes, n))
	}

	c := len(partialRes)
	for r := 0; r < n; r++ {
		_, ok := rows[r]
		_, ok2 := diags[r-c]
		_, ok3 := antiDiags[r+c]
		if !ok && !ok2 && !ok3 {
			rows[r], diags[r-c], antiDiags[r+c] = true, true, true
			partialRes = append(partialRes, r)
			backtrack(rows, diags, antiDiags, n, partialRes, res)
			partialRes = partialRes[:len(partialRes)-1]
			delete(rows, r)
			delete(diags, r-c)
			delete(antiDiags, r+c)
		}
	}
}

func writeOutNQueens(arr []int, n int) []string {
	res := make([]string, 0, n)

	for _, el := range arr {
		res = append(res, strings.Repeat(".", el)+"Q"+strings.Repeat(".", n-el-1))
	}
	return res
}

File: python/n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        def backtrack(rows, diags, anti_diags, partial_res, res):
            if len(partial_res) >= n:
                res.append(write_out_nqueens(partial_res))
                return
            
            c = len(partial_res)
            for r in range(n):
                if r not in rows and r-c not in diags and r+c not in anti_diags:
                    rows.add(r); diags.add(r-c); anti_diags.add(r+c)
                    partial_res.append(r)
                    backtrack(rows, diags, anti_diags, partial_res, res)
                    partial_res.pop()
                    rows.remove(r); diags.remove(r-c); anti_diags.remove(r+c)
        def write_out_nqueens(arr):
            res = []
            for el in arr:
                res.append("."*el + "Q" + "."*(n-el-1))
            return res
        
        res = []
        backtrack(set(), set(), set(), [], res)
        return res